draft/celery-5-high-level-architecture.rst

```diff
@@ -0,0 +1,1216 @@
+
+Publisher Circuit Breakers
+++++++++++++++++++++++++++
+
+Each :ref:`health check <draft/celery-5-high-level-architecture:Health Checks>`
+has it's own Circuit Breaker.
+Once a circuit breaker is tripped, the :term:`messages <Message>` are stored
+in the :ref:`draft/celery-5-high-level-architecture:messages backlog` until
+the health check recovers and the circuit is once again closed.
+
+Messages Backlog
+++++++++++++++++
+
+The messages backlog is a temporary queue of :term:`messages <Message>`
+yet to be published to the appropriate broker cluster.
+
+In the event where :term:`messages <Message>` cannot be published
+for any reason, the :term:`messages <Message>` are kept inside the queue.
+
+By default, an in-memory queue will be used. The user may provide another
+implementation which stores the :term:`messages <Message>` on-disk
+or in a central database.
+
+Implementers should take into account what happens whenever writing to the
+messages backlog fails.
+
+The default fallback mechanism will append the :term:`messages <Message>` into
+an in-memory queue.
+These :term:`messages <Message>` will be published first in order to avoid
+:term:`Message` loss in case the publisher goes down for any reason.
+
+Publisher Daemon
+++++++++++++++++
+
+In sufficiently large deployments, one server runs multiple workloads which
+may publish to a :term:`Message Broker`.
+
+Therefore, it is unnecessary to maintain a publisher for each process that
+publishes to a :term:`Message Broker`.
+
+In such cases, a Publisher Daemon can be used. The publishing processes will
+specify it as their target and communicate the :term:`messages <Message>`
+to be published via a socket.
+
+If a disk based queue is used, the user may configure Celery to write to it
+directly, provided that the queue can perform inserts and deletes concurrently.
+
+Router
+------
+
+The Router is responsible for managing the connection to a :term:`message broker`
+and consuming :term:`messages <Message>` from the :term:`message broker`.
+
+The Router can maintain a connection to a cluster of
+:term:`message brokers <message broker>` or even clusters of
+:term:`message brokers <message broker>`.
+
+Data Source
++++++++++++
+
+Ingress Only Data Sources
+~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Ingress/Egress Data Sources
+~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Controller
+----------
+
+The Controller is responsible for managing the lifecycle of all other Celery
+components.
+
+It spawns the :ref:`Workers <draft/celery-5-high-level-architecture:Worker>`, :ref:`Routers <draft/celery-5-high-level-architecture:Router>`,
+:ref:`Schedulers <draft/celery-5-high-level-architecture:Scheduler>` and if configured and possible,
+the :term:`Message Brokers <Message Broker>` as well.
+
+By default, the Controller creates sub-processes for
+all the required components. This is suitable for small scale deployments
+or for deployments where SystemD is unavailable.
+
+Foreman
```

Comment:
Perhaps a reference to [supervisord](http://supervisord.org) would also be useful, as I think it is more popular and has more features than Foreman. Also, Heroku seems to have migrated to [Forego](https://devcenter.heroku.com/changelog-items/692).

Reply:
The Foreman is simply the name of the service. I haven't worked a lot on that section yet.